src/sources/dnstap/mod.rs

```diff
@@ -246,12 +332,83 @@ impl FrameHandler for DnstapFrameHandler {
+    fn host_key(&self) -> &str {
+        self.host_key.as_str()
+    }
+
+    fn source_type_key(&self) -> &str {
+        self.source_type_key.as_str()
+    }
+
+    fn timestamp_key(&self) -> &str {
+        self.timestamp_key.as_str()
     }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
 
-    fn timestamp_key(&self) -> String {
-        self.timestamp_key.clone()
+    #[test]
+    fn simple_matches_schema() {
+        let record = r#"{"dataType":"Message",
+                         "dataTypeId":1,
+                         "messageType":"ClientQuery",
+                         "messageTypeId":5,
+                         "requestData":{
+                           "fullRcode":0,
+                           "header":{
+                             "aa":false,
+                             "ad":true,
+                             "anCount":0,
+                             "arCount":1,
+                             "cd":false,
+                             "id":38339,
+                             "nsCount":0,
+                             "opcode":0,
+                             "qdCount":1,
+                             "qr":0,
+                             "ra":false,
+                             "rcode":0,
+                             "rd":true,
+                             "tc":false},
+                           "opt":{
+                             "do":false,
+                             "ednsVersion":0,
+                             "extendedRcode":0,
+                             "options":[{"optCode":10,
+                                         "optName":"Cookie",
+                                         "optValue":"5JiWq4VYa7U="}],
+                             "udpPayloadSize":1232},
+                           "question":[{"class":"IN","domainName":"whoami.example.org.","questionType":"A","questionTypeId":1}],
+                           "rcodeName":"NoError",
+                           "time":1667909880863224758,
+                           "timePrecision":"ns"},
+                         "serverId":"stephenwakely-Precision-5570",
+                         "serverVersion":"CoreDNS-1.10.0",
+                         "socketFamily":"INET",
+                         "socketProtocol":"UDP",
+                         "sourceAddress":"0.0.0.0",
+                         "sourcePort":54782,
+                         "source_type":"dnstap",
+                         "time":1667909880863224758,
+                         "timePrecision":"ns"
+                         }"#;
+
+        let json: serde_json::Value = serde_json::from_str(record).unwrap();
+        let mut event = Event::from(LogEvent::from(value::Value::from(json)));
+        event.as_mut_log().insert("timestamp", chrono::Utc::now());
+
+        let definition = DnstapConfig::event_schema("timestamp");
+        let schema = vector_core::schema::Definition::empty_legacy_namespace()
+            .with_standard_vector_source_metadata();
+
+        assert_eq!(
```

Comment:
you can use `definition.assert_valid_for_event(&event))` which will do the assertion for you, and the error will print a bit better